### PR TITLE
 feature - add ctrl click functionality to drill down to referenced records

### PR DESCRIPTION
**Pull request by schofida**
RecordTreeView:
Added new css and 4 new events to detect mouseover/mouseout and keydown/keyup hotkey events. Detects if the ctrl key is pressed and if the pointer is over an element that links to another record. On ctrl+click, drill down into the selected record.

Hotkey functionality:
Added capability to specify keyup callbacks. Keyup configurations will be designated with the 'Up' suffix (e.g BaseHotkeysUp)

NOTE: I noticed the action.callback that you added (I did the same, sorry I never had time to tell you of it). However the action.Constructor === function change was not detecting functions properly (see my Ctrl callbacks in addition to the Ctrl+S callback) and would cause an error. I reverted that line back to typeStr === 'function' which appears to be working.

NOTE 2: In addition of me being new to Angular, I am also apparently a total dink when it comes to Git. Hopefully I'm getting this pull request thing to you properly and that this proper protocol for getting these changes to you. Let me know if there is anything that I would need to do differently.

### CODE REVIEW

**Comment by matortheeternal** [comment]
Will do code review soon.  :coffe:

File: src/javascripts/Services/hotkeyService.js [hunk 1]
```diff
@@ -33,19 +33,30 @@ ngapp.service('hotkeyService', function(hotkeyFactory) {
...
             if (!hotkeys[hotkey]) return;
             trigger(scope, hotkeys[hotkey], e);
-        };
+        }
+    }
+    
+    this.buildOnKeyDown = function(scope, label, view) {
+        let hotkeys = hotkeyFactory[`${view}Hotkeys`];
+        scope[label] = keyEventHandler(scope, hotkeys,'keydown');
+    };
```
Comment (matortheeternal) [inline src/javascripts/Services/hotkeyService.js]: 🚀 Need a new line between these functions.

File: src/javascripts/Services/hotkeyService.js [hunk 2]
```diff
@@ -33,19 +33,30 @@ ngapp.service('hotkeyService', function(hotkeyFactory) {
...
         e.stopImmediatePropagation();
         e.preventDefault();
     };
 
-    this.buildOnKeyDown = function(scope, label, view) {
-        let hotkeys = hotkeyFactory[`${view}Hotkeys`];
-        scope[label] = function(e) {
+    let keyEventHandler = function(scope,hotkeys,type) {
+        return function (e) {
+            if (e.type !== type)
```
Comment (matortheeternal) [inline src/javascripts/Services/hotkeyService.js]: let's make this a one-line if statement to match the style used in the rest of the codebase.  so:
```js
if (e.type !== type) return;
```

File: src/javascripts/Services/hotkeyService.js [hunk 3]
```diff
@@ -33,19 +33,30 @@ ngapp.service('hotkeyService', function(hotkeyFactory) {
...
             }) || 'else';
             if (!hotkeys[hotkey]) return;
             trigger(scope, hotkeys[hotkey], e);
-        };
+        }
+    }
+    
+    this.buildOnKeyDown = function(scope, label, view) {
+        let hotkeys = hotkeyFactory[`${view}Hotkeys`];
+        scope[label] = keyEventHandler(scope, hotkeys,'keydown');
```
Comment (matortheeternal) [inline src/javascripts/Services/hotkeyService.js]: :rocket: need a space between `,` and `'keydown'`

File: src/javascripts/Services/hotkeyService.js [hunk 4]
```diff
@@ -33,19 +33,30 @@ ngapp.service('hotkeyService', function(hotkeyFactory) {
...
+    };
+    this.buildOnKeyUp = function(scope, label, view) {
+        let hotkeys = hotkeyFactory[`${view}HotkeysUp`];
+        scope[label] = keyEventHandler(scope, hotkeys,'keyup');
```
Comment (matortheeternal) [inline src/javascripts/Services/hotkeyService.js]: 🚀 need a space between `,` and `'keyup'`

File: src/javascripts/Services/hotkeyService.js [hunk 5]
```diff
@@ -33,19 +33,30 @@ ngapp.service('hotkeyService', function(hotkeyFactory) {
...
+                return;
+            
             let hotkey = Object.keys(hotkeys).find(function(key) {
                 return e.keyCode === keycodes[key];
             }) || 'else';
             if (!hotkeys[hotkey]) return;
             trigger(scope, hotkeys[hotkey], e);
-        };
+        }
+    }
```
Comment (matortheeternal) [inline src/javascripts/Services/hotkeyService.js]: 🐤 should have a semicolon terminating this line

File: src/javascripts/Views/recordTreeView.js [hunk 6]
```diff
@@ -52,12 +52,47 @@ ngapp.controller('recordTreeViewController', function($scope, $element, $timeout
         e.stopImmediatePropagation();
     };
 
-    $scope.onCellMouseDown = function(index) {
+    $scope.onCellMouseDown = function(e, node, index) {
+    
+        if (e.ctrlKey && index > 0 && node.value_type === xelib.vtReference)
+        {
```
Comment (matortheeternal) [inline src/javascripts/Views/recordTreeView.js]: 🐤 this brace should be on the previous line.

File: src/javascripts/Views/recordTreeView.js [hunk 7]
```diff
@@ -52,12 +52,47 @@ ngapp.controller('recordTreeViewController', function($scope, $element, $timeout
         e.stopImmediatePropagation();
     };
 
-    $scope.onCellMouseDown = function(index) {
+    $scope.onCellMouseDown = function(e, node, index) {
+
```
Comment (matortheeternal) [inline src/javascripts/Views/recordTreeView.js]: 👾 this newline is unnecessary.  space invaders will invade other unnecessary spaces.

File: src/javascripts/Views/recordTreeView.js [hunk 8]
```diff
@@ -52,12 +52,47 @@ ngapp.controller('recordTreeViewController', function($scope, $element, $timeout
...
+            index -= 1;
+            const id = $scope.getRecord(index);
+            const path = xelib.LocalPath(node.first_handle);
+            const ref = xelib.GetLinksTo(id,path);
+            
+            if (ref > 0) {
+                $scope.record = ref;
+                return;
+            }
+
```
Comment (matortheeternal) [inline src/javascripts/Views/recordTreeView.js]: 👾

File: src/javascripts/Views/recordTreeView.js [hunk 9]
```diff
@@ -52,12 +52,47 @@ ngapp.controller('recordTreeViewController', function($scope, $element, $timeout
...
+            }
+            
+        }
         let oldIndex = $scope.focusedIndex;
         $scope.focusedIndex = index;
         if (oldIndex !== index) $timeout($scope.updateNodeLabels);
     };
 
+    $scope.onCellMouseOver = function(e, node, index) {
+
```
Comment (matortheeternal) [inline src/javascripts/Views/recordTreeView.js]: 👾

File: src/javascripts/Views/recordTreeView.js [hunk 10]
```diff
@@ -52,12 +52,47 @@ ngapp.controller('recordTreeViewController', function($scope, $element, $timeout
...
+        if (e.srcElement && e.ctrlKey && index > 0 && 
+            node.value_type === xelib.vtReference)
+        {
```
Comment (matortheeternal) [inline src/javascripts/Views/recordTreeView.js]: 🐤 this brace should be on the previous line.

File: src/javascripts/Views/recordTreeView.js [hunk 11]
```diff
@@ -52,12 +52,47 @@ ngapp.controller('recordTreeViewController', function($scope, $element, $timeout
...
+            e.srcElement.classList.add('highlight-reference');
+        }
+        //Helper variables for on controlkeypressed listener
+        $scope.highlightedCell = e.srcElement;
+        $scope.highlightedNode = node;
+    };
+    
+    $scope.onCellMouseLeave = function(e, node, index) {
+
```
Comment (matortheeternal) [inline src/javascripts/Views/recordTreeView.js]: 👾

File: src/javascripts/Views/recordTreeView.js [hunk 12]
```diff
@@ -98,6 +133,22 @@ ngapp.controller('recordTreeViewController', function($scope, $element, $timeout
         if (!xelib.GetFormID($scope.record)) $scope.reload();
     });
 
+    //Event broadcasts from baseHotKey factory
+    $scope.$on('controlKeyPressed', function(){
+        if ($scope.highlightedCell && $scope.highlightedNode && 
+            $scope.highlightedNode.value_type === xelib.vtReference &&
+            !$scope.highlightedCell.classList.contains('highlight-reference'))
+        {
```
Comment (matortheeternal) [inline src/javascripts/Views/recordTreeView.js]: 🐤 this brace should be on the previous line.

File: src/javascripts/Views/recordTreeView.js [hunk 13]
```diff
@@ -52,12 +52,47 @@ ngapp.controller('recordTreeViewController', function($scope, $element, $timeout
         e.stopImmediatePropagation();
     };
 
-    $scope.onCellMouseDown = function(index) {
+    $scope.onCellMouseDown = function(e, node, index) {
+    
+        if (e.ctrlKey && index > 0 && node.value_type === xelib.vtReference)
+        {
+            index -= 1;
```
Comment (matortheeternal) [inline src/javascripts/Views/recordTreeView.js]: index is only being used once in this code block, so you should just do `index - 1` on the next line

File: src/javascripts/Views/recordTreeView.js [hunk 14]
```diff
@@ -52,12 +52,47 @@ ngapp.controller('recordTreeViewController', function($scope, $element, $timeout
...
+            const id = $scope.getRecord(index);
+            const path = xelib.LocalPath(node.first_handle);
+            const ref = xelib.GetLinksTo(id,path);
```
Comment (matortheeternal) [inline src/javascripts/Views/recordTreeView.js]: 🚀 need a space between `id,` and `path`

File: src/javascripts/Services/hotkeyService.js [hunk 15]
```diff
@@ -33,19 +33,30 @@ ngapp.service('hotkeyService', function(hotkeyFactory) {
...
-        action.Constructor === Function ? action(scope, e) : scope[action](e);
+        typeStr === 'function' ? action(scope, e) : scope[action](e);
         e.stopImmediatePropagation();
         e.preventDefault();
     };
 
-    this.buildOnKeyDown = function(scope, label, view) {
-        let hotkeys = hotkeyFactory[`${view}Hotkeys`];
-        scope[label] = function(e) {
+    let keyEventHandler = function(scope,hotkeys,type) {
```
Comment (matortheeternal) [inline src/javascripts/Services/hotkeyService.js]: :rocket: need a space between `scope,` and `hotkeys`.  also need a space between `hotkeys,` and `type`

**Comment by schofida** [comment]
I push the formatting changes and it looks like it was added to this pull request. But again, still new to Git. Let me know if you need anything else.

**Pull request merged**